packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_prop/string_prop.rs

```diff
@@ -72,20 +123,39 @@ impl PropUpdater<String, RequiredData> for StringProp {
         self.default_value.clone()
     }
 
-    fn return_data_queries(&self) -> Vec<Option<DataQuery>> {
+    fn return_data_queries(&self) -> Vec<DataQuery> {
         RequiredDataQueries {
-            strings: Some(self.data_query.clone()),
+            independent_state: DataQuery::State,
+            strings: self.data_query.clone(),
         }
         .into()
     }
 
     fn calculate(&mut self, data: &RequiredData) -> PropCalcResult<String> {
         match data.strings.len() {
-            0 => PropCalcResult::Calculated(String::from("")),
+            0 => {
+                if self.propagate_came_from_default {
+                    // if propagate_came_from_default is true,
+                    // then always_return_value is true on the string data query,
+                    // so we should never reach this
+                    unreachable!()
+                }
+
+                // If propagate_came_from_default is false and there were no dependencies returned,
+                // then use the independent state, propagating its came_from_default as well as its value.
+                // In this way, came_from_default will be true only if no dependencies were returned.
+                data.independent_state.prop_calc_result()
```

Comment:
What's the difference between `prop_calc_result()` and `calc()`? Can the functions have the same name?

Reply:
I'm not sure, as we don't have a `calc`. `prop_calc_result()` propagates both the value and the `came_from_default` of the prop.

Reply:
Not sure the best name for it. It returns a `PropCalcResult` based on the prop. If it creates confusion, I could just remove it and write the `if` statement manually each time.

At one point it was instead
`PropCalcResult::From(data.indepenent_state)`
which sent a reference to `data.independent_state`. But sending a reference is no longer an option due to Rust borrowing rules and the fact that I now mutate `data.independent_state` immediately after this function returns to mark it viewed and before processing the `PropCalcResult`.

Reply:
To make the propagation of `came_from_default` more explicit, I'm removing this function and writing the `if` statement manually each time.